src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java21/InputStringTemplateBasic.java

```diff
@@ -0,0 +1,49 @@
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.grammar.java21;
+
+public class InputStringTemplateBasic {
+    int x = 10;
+    int y = 20;
+
+    // most basic example, empty string template expression
+    String result1 = STR."";
+
+    // simple single value interpolation
+    String result2 = STR. "x\{ x }x";
```

Comment:
```
        |--VARIABLE_DEF -> VARIABLE_DEF [12:4]
        |   |--MODIFIERS -> MODIFIERS [12:4]
        |   |--TYPE -> TYPE [12:4]
        |   |   `--IDENT -> String [12:4]
        |   |--IDENT -> result2 [12:11]
        |   |--ASSIGN -> = [12:19]
        |   |   `--EXPR -> EXPR [12:24]
        |   |       `--DOT -> . [12:24]
        |   |           |--IDENT -> STR [12:21]
        |   |           `--STRING_TEMPLATE_BEGIN -> "x\{ [12:26]
        |   |               |--EMBEDDED_EXPRESSION -> EMBEDDED_EXPRESSION [12:31]
        |   |               |   `--IDENT -> x [12:31]
        |   |               `--STRING_TEMPLATE_END -> }x" [12:33]
        |   `--SEMI -> ; [12:36]
```

Reply:
Need to add a case with no embedded expression (like no x here)

Reply:
in my imagination, it bit more trrucrured:
```
`--STRING_TEMPLATE_BEGIN -> " [12:26]
        |   |               |--TEMPLATE_CONTENT "x"
        |   |               |--EMBEDED_BEGIN -> "\{
        |   |               |--EMBEDDED_EXPRESSION -> EMBEDDED_EXPRESSION [12:31]
        |   |               |   `--IDENT -> x [12:31]
        |   |               |--EMBEDED_END -> }
        |   |               |--TEMPLATE_CONTENT--x" [12:33]
        |   |               `--STRING_TEMPLATE_END -> " [12:33]
```

Reply:
I also prefer your proposal, however this will be a bit more challenging to create. Let me see what I can do. However, the current implementation is more reflective of the actual java grammar.

Reply:
Ok:
```
STRING_TEMPLATE_BEGIN -> " [12:26]
|--STRING_TEMPLATE_CONTENT -> x [12:27]
|--EMBEDDED_EXPRESSION_BEGIN -> \{ [12:28]
|--EMBEDDED_EXPRESSION -> EMBEDDED_EXPRESSION [12:31]
|   `--IDENT -> x [12:31]
|--EMBEDDED_EXPRESSION_END -> } [12:33]
|--STRING_TEMPLATE_CONTENT -> x [12:34]
`--STRING_TEMPLATE_END -> " [12:35]
```

Reply:
Looks good.

Reply:
This is the new AST for this code, all other examples in this PR have been edited:
```
        |--VARIABLE_DEF -> VARIABLE_DEF [12:4]
        |   |--MODIFIERS -> MODIFIERS [12:4]
        |   |--TYPE -> TYPE [12:4]
        |   |   `--IDENT -> String [12:4]
        |   |--IDENT -> x [12:11]
        |   |--ASSIGN -> = [12:13]
        |   |   `--EXPR -> EXPR [12:18]
        |   |       `--DOT -> . [12:18]
        |   |           |--IDENT -> STR [12:15]
        |   |           `--STRING_TEMPLATE_BEGIN -> " [12:19]
        |   |               |--STRING_TEMPLATE_CONTENT ->  [12:20]
        |   |               |--EMBEDDED_EXPRESSION_BEGIN -> \{ [12:20]
        |   |               |--EMBEDDED_EXPRESSION -> EMBEDDED_EXPRESSION [12:24]
        |   |               |--EMBEDDED_EXPRESSION_END -> } [12:24]
        |   |               |--STRING_TEMPLATE_CONTENT ->  [12:25]
        |   |               `--STRING_TEMPLATE_END -> " [12:25]
        |   `--SEMI -> ; [12:26]


```

Reply:
All other examples have been edited, this is the new AST for this code (it has moved down in the source code file):
```
        |--VARIABLE_DEF -> VARIABLE_DEF [15:4]
        |   |--MODIFIERS -> MODIFIERS [15:4]
        |   |--TYPE -> TYPE [15:4]
        |   |   `--IDENT -> String [15:4]
        |   |--IDENT -> result2 [15:11]
        |   |--ASSIGN -> = [15:19]
        |   |   `--EXPR -> EXPR [15:24]
        |   |       `--DOT -> . [15:24]
        |   |           |--IDENT -> STR [15:21]
        |   |           `--STRING_TEMPLATE_BEGIN -> " [15:26]
        |   |               |--STRING_TEMPLATE_CONTENT -> x [15:27]
        |   |               |--EMBEDDED_EXPRESSION_BEGIN -> \{ [15:28]
        |   |               |--EMBEDDED_EXPRESSION -> EMBEDDED_EXPRESSION [15:31]
        |   |               |   `--IDENT -> x [15:31]
        |   |               |--EMBEDDED_EXPRESSION_END -> } [15:33]
        |   |               |--STRING_TEMPLATE_CONTENT -> x [15:34]
        |   |               `--STRING_TEMPLATE_END -> " [15:35]
        |   `--SEMI -> ; [15:36]


```